### PR TITLE
Rewrite tests to make use of `pytest`

### LINKED ISSUE #37: Implement tests using `Pytest`
Tests are currently executed at the end of each step (at runtime) to validate outputs. It would be preferred to also have tests that run separately. These generally fall into two categories:

- unit tests (test small components - i.e. functions)
- integration tests (test the execution of many components together)

The recommended testing framework is `Pytest`. Create a `tests` directory and run tests automatically when PRs are issued using github workflows.

### PR DESCRIPTION
**Pull request by tztsai**
This PR rewrites the testing approach in SPINacc. Previously, tests were written to execute at the end of a run. There are two problems: 

- This lengthens the program execution time
- Tests have only been written for a single version of ORCHIDEE (corresponding to the data on Zenodo). 

Instead, this PR introduces `pytest`, which allows users to run tests after a step of the pipeline has completed. To utilise `pytest` all previous tests have been renamed to `test_task*.py` and all functions now have a `test` prefix. 

To summarise:

- Renamed `config.py` to `conftest.py` (a`pytest` recognised configuration file)
- Renamed all previous `tests` files to `test_task*.py` and all test functions renamed with a prefix `test_`, again this is required naming convention for `pytest`.
- Pytest markers are used to parametrize the input variables such as the names of the files to be compared.
- Constant variables used in the testing are defined as pytest fixtures,.
- Commonly used functions that compare the contents of the reference file and the result file have been moved into `utils.py`. 
- The `utils.py` includes a recursive compare function, as this was necessary to check arrays of arrays. 

### Execution
To run the tests, use the command `pytest tests/`.

In order to merge the following tasks must be completed:

- [x] Address Tom's PR comments (add docs, clarify code, fix version)
- [x] Reference data mismatch with `IDx.npy` caused by unique call, see below. Subsequent code is correct, but this needs to be confirmed with a test. Perhaps reference data needs updating. Try pipeline without unique in `Cluster.py` first. 
- [x] `IDloc.npy` and `IDsel.npy` now same as reference (not critical as only used for plotting) 
- [x] Fixed discrepancy between *.txt files and `MLacc_results.csv`. 

Documentation on running tests will be provided in a later PR. 

Closes #37 and #40.

### CODE REVIEW

**Comment by ma595** [comment]
All tests pass locally for me, except for step 3 which we don't care about at the moment.

File: tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import numpy as np
+import pytest
+
+
+@pytest.fixture
+def reference_path():  # reference EXE_DIR in zenodo
+    return "/home/surface10/mrasolon/files_for_zenodo/reference/EXE_DIR/"
```
Comment (ma595) [inline tests/conftest.py:7]: Is there a way to pass this as a parameter, or have it as some pytest config file?

File: tests/utils.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        if arr1.dtype == "object" and arr2.dtype == "object":
+            for sub_arr1, sub_arr2 in zip(arr1, arr2):
+                if not recursive_compare(sub_arr1, sub_arr2, tol):
+                    return False
+            return True
+        else:
+            # Compare the arrays element-wise for floating point numbers
+            return np.allclose(arr1, arr2, atol=tol)
+    else:
+        return False
```
Comment (ma595) [inline tests/utils.py:31]: I've added a recursive compare function that compares ndarrays of ndarrays...

**Review by ma595 — CHANGES_REQUESTED** [review]
This is a good start on regression testing, however we need to think about automated testing, and given that some of the steps have rather large data dependencies, I'm struggling to see how this can be done.

File: tests/conftest.py (hunk 1 above)
Comment (tztsai) [reply tests/conftest.py:7]: Yeah I can add the config path as an argument, and return the reference data path in the config

**Review by ma595 — APPROVED** [review]

File: requirements.txt [hunk 3]
```diff
@@ -7,3 +7,4 @@ pandas==1.5.2
 scikit_learn==1.1.3
 scipy==1.9.3
 xarray==2022.11.0
+pytest
```
Comment (TomMelt) [inline requirements.txt:10]: Can we fix or bound the `pytest` version?
```suggestion
pytest>=1.0.0
```

File: tests/conftest.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import numpy as np
+import pytest
+
+
+config_path = "DEF_Trunk/MLacc.def"
+
+
+@pytest.fixture
+def reference_path():  # reference EXE_DIR in zenodo
```
Comment (TomMelt) [inline tests/conftest.py:9]: Can we add docstrings?

File: tests/test_task1.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+from utils import compare_npy_files
+
+
+@pytest.mark.parametrize(
+    "filename",
+    [
+        "dist_all.npy"
+    ],  # "auxil.npy" & "packdata.npy" have been replaced by "packdata.nc"
+)
+def test_compare_npy_files(reference_path, test_path, filename):
```
Comment (TomMelt) [inline tests/test_task1.py:11]: Like before, can we add docstring to all the tests?

File: tests/test_task4.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    files1 = [
+        f
+        for f in os.listdir(reference_path)
+        if os.path.isfile(os.path.join(reference_path, f))
+    ]
+    files2 = [
+        f for f in os.listdir(test_path) if os.path.isfile(os.path.join(test_path, f))
+    ]
+
+    common_files = set(files1) & set(files2)
```
Comment (TomMelt) [inline tests/test_task4.py:17]: what does this line of code mean?

**Review by TomMelt — CHANGES_REQUESTED** [review]
Looks good. Just a couple of outstanding questions.

**Comment by ma595** [comment]
@ma595 needs to address Tom's comments above.

**Comment by ma595** [comment]
There is a mismatch caused by the L129 in `Cluster.py` `IDx = np.unique(IDx, axis=0)` in commit `df4e9d5`. 

This reduces the `IDx.npy` from length 693 (the original length) to 586 (the new length). 

Therefore, all new code is correct. I will write a test reading the reference IDx and taking unique.

File: tests/test_task4.py (hunk 6 above)
Comment (ma595) [reply tests/test_task4.py:17]: This test has been skipped - we need to establish how to properly test the ML performance.

**Comment by ma595** [comment]
Summarising progress on this PR:

I've added a check of the `packdata.nc` and have fixed mismatches between the reference and test output in the `Cluster.py` file. For the moment, I have commented out the `np.unique` call - I will fix this to ensure that tests still pass. The reference data is now incorrect.

**Comment by ma595** [comment]
Discrepancies between `MLacc_results.txt` and reference files (I've checked out `44b7a5b` as a check). 

```
Some values in R2 do not match MLacc_results.csv
   comp        var        R2 comp  var    R2
8   som  carbon_04  0.934900  som    3  0.92
33  som  carbon_13  0.959239  som   12  0.97
All values in dNRMSE match MLacc_results.csv
Some values in slope do not match MLacc_results.csv
   comp        var     slope comp  var  slope
7   som  carbon_04  1.017391  som    3   1.03
9   som  carbon_05  0.910774  som    4   0.89
10  som  carbon_05  0.921600  som    4   0.90
11  som  carbon_05  0.945176  som    4   0.93
15  som  carbon_07  0.918745  som    6   0.93
16  som  carbon_07  0.939137  som    6   0.96
30  som  carbon_12  0.974204  som   11   0.99
31  som  carbon_12  0.997260  som   11   1.01
33  som  carbon_13  0.947914  som   12   0.98
36  som  carbon_14  0.965142  som   13   0.94
40  som  carbon_15  0.967546  som   14   0.98
Some values in sNRMSE do not match MLacc_results.csv
   comp        var    sNRMSE comp  var  sNRMSE
7   som  carbon_04  0.248488  som    3    0.27
8   som  carbon_04  0.255146  som    3    0.28
9   som  carbon_05  0.228092  som    4    0.25
10  som  carbon_05  0.211208  som    4    0.23
17  som  carbon_07  0.193161  som    6    0.18
26  som  carbon_10  0.160929  som    9    0.15
33  som  carbon_13  0.201893  som   12    0.16
34  som  carbon_13  0.161907  som   12    0.14
All values in iNRMSE match MLacc_results.csv
Some values in f_SDSD do not match MLacc_results.csv
   comp        var    f_SDSD comp  var  f_SDSD
7   som  carbon_04  0.035841  som    3    0.06
9   som  carbon_05  0.082083  som    4    0.11
10  som  carbon_05  0.074819  som    4    0.11
11  som  carbon_05  0.017909  som    4    0.05
15  som  carbon_07  0.111631  som    6    0.09
16  som  carbon_07  0.050525  som    6    0.02
19  som  carbon_08  0.048112  som    7    0.03
20  som  carbon_08  0.057761  som    7    0.03
21  som  carbon_09  0.118936  som    8    0.14
22  som  carbon_09  0.102351  som    8    0.09
27  som  carbon_11  0.024329  som   10    0.01
30  som  carbon_12  0.013586  som   11    0.00
31  som  carbon_12  0.002480  som   11    0.02
33  som  carbon_13  0.025480  som   12    0.00
40  som  carbon_15  0.015320  som   14    0.00
All values in f_SB match MLacc_results.csv
Some values in f_LSC do not match MLacc_results.csv
   comp        var     f_LSC comp  var  f_LSC
7   som  carbon_04  0.960445  som    3   0.94
9   som  carbon_05  0.917887  som    4   0.89
10  som  carbon_05  0.924856  som    4   0.89
11  som  carbon_05  0.980707  som    4   0.95
15  som  carbon_07  0.872242  som    6   0.90
16  som  carbon_07  0.941257  som    6   0.98
19  som  carbon_08  0.949659  som    7   0.97
20  som  carbon_08  0.937724  som    7   0.97
21  som  carbon_09  0.863930  som    8   0.84
27  som  carbon_11  0.975425  som   10   0.99
30  som  carbon_12  0.986397  som   11   1.00
31  som  carbon_12  0.992699  som   11   0.98
33  som  carbon_13  0.974414  som   12   1.00
40  som  carbon_15  0.983023  som   14   1.00
.
```

**Comment by ma595** [comment]
We need to reassign the random seed for step 4. Results are now the same regardless.

**Comment by tztsai** [comment]
Thanks for figuring out the cause!

**Comment by ma595** [comment]
Hi @tztsai, could you please review this. Thanks.

File: tests/test_task3.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import pytest
+from utils import compare_nc_files
+
+
+@pytest.fixture
+def file_prefix():
+    return "forcing_aligned_"
+
+
+@pytest.mark.skip("Skipped for redundancy of tests.")
```
Comment (tztsai) [inline tests/test_task3.py:10]: Are these tests unnecessary any more?

File: Tools/Cluster.py [hunk 8]
```diff
@@ -126,7 +125,10 @@ def Cluster_all(packdata, varlist, KK, logfile):
 
     # 5. Output the ID
     IDx = np.concatenate([adict["PFT%itrainingID" % ii] for ii in kpfts])
-    IDx = np.unique(IDx, axis=0)
+
+    # 6. If take_unique is False, we are equivalent to the original implementation
+    if take_unique:
+        IDx = np.unique(IDx, axis=0)
```
[the remark this answers is not in the document]
Comment (ma595) [reply Tools/Cluster.py:131]: It does.

File: tests/test_task3.py (hunk 7 above)
Comment (ma595) [reply tests/test_task3.py:10]: Yes, unnecessary.

File: Tools/Cluster.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (tztsai) [reply Tools/Cluster.py:131]: For better or for worse?

[the remark this answers is not in the document]
Comment (ma595) [reply Tools/Cluster.py:131]: I'll rerun and compare. My feeling is that taking `np.unique` gives better results.

[the remark this answers is not in the document]
Comment (ma595) [reply Tools/Cluster.py:131]: `len(IDx) = 693` using `unique = False` (This corresponds to the original). 
`len(IDx) = 586` using `unique = True`

We are approximately the same up to the 2nd / 3rd s.f on R2. Hard to say whether taking unique is better generally.

[the remark this answers is not in the document]
Comment (ma595) [reply Tools/Cluster.py:131]: Will leave this investigation to future work if time allows.

**Comment by ma595** [comment]
I rebased off of fix-evaluation - this error is due to the sorting of the indices. After correcting. we get the following discrepancies: 

Some values in R2 do not match MLacc_results.csv
   comp           var        R2    comp  var    R2
2   som  litter_03_ab  0.680444  litter    2  0.71
7   som  litter_05_ab  0.937575  litter    4  0.95
24  som  litter_14_ab  0.518584  litter   13  0.53
30  som  litter_03_be  0.651914  litter   16  0.59
35  som  litter_05_be  0.833144  litter   18  0.85
36  som  litter_06_be  0.926984  litter   19  0.90
44  som  litter_10_be  0.919111  litter   23  0.93
46  som  litter_11_be  0.621841  litter   24  0.73

**Comment by ma595** [comment]
After the previous commit all tests now pass and this is ready to merge. Possibly consider adding a test to step 5, given that visualisation now works following #84

**Pull request merged**